server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -241,139 +239,131 @@ synchronized void startWatcher(ClusterState clusterState, boolean onStartup) {
+                 *     ENTRY_MODIFY:operator
+                 * After we get an indication that something has changed, we check the timestamp, file id,
+                 * real path of our desired file. We don't actually care what changed, we just re-check ourselves.
+                 */
+                if (Files.exists(operatorSettingsDir)) {
+                    try {
+                        Path path = operatorSettingsFile();
+
+                        var events = key.pollEvents();
+                        if (logger.isDebugEnabled()) {
+                            events.forEach(e -> logger.debug("{}:{}", e.kind().toString(), e.context().toString()));
                         }
-                    } else {
-                        key.pollEvents();
                         key.reset();
+
+                        // We re-register the settings directory watch key, because we don't know
+                        // if the file name maps to the same native file system file id. Symlinks
+                        // are one potential cause of inconsistency here, since their handling by
+                        // the WatchService is platform dependent.
+                        settingsDirWatchKey = enableSettingsWatcher(settingsDirWatchKey, operatorSettingsDir);
+
+                        if (watchedFileChanged(path)) {
+                            try {
+                                processFileSettings(path).get();
+                            } catch (ExecutionException e) {
+                                logger.error("Error processing operator settings json file", e.getCause());
+                            }
+                        }
+                    } catch (IOException e) {
+                        logger.warn("encountered I/O error while watching file settings", e);
                     }
+                } else {
+                    key.pollEvents();
+                    key.reset();
                 }
-            } catch (ClosedWatchServiceException | InterruptedException expected) {
-                logger.info("shutting down watcher thread");
-            } catch (Exception e) {
-                logger.error("shutting down watcher thread with exception", e);
-            } finally {
-                watcherThreadLatch.countDown();
             }
-        }, "elasticsearch[file-settings-watcher]").start();
+        } catch (ClosedWatchServiceException | InterruptedException expected) {
+            logger.info("shutting down watcher thread");
+        } catch (Exception e) {
+            logger.error("shutting down watcher thread with exception", e);
+        }
     }
 
     synchronized void stopWatcher() {
         logger.debug("stopping watcher ...");
         if (watching()) {
-            try {
-                // make sure the watcher thread hits the processing latch correctly
+            try (var ws = watchService) {   // make sure watch service is closed whatever
+                watcherThread.interrupt();
```

Comment:
What if we fail with interrupted exception here? Or on any exception in the join? I think we should move the clean-up of the watch keys in finally, now that it's not critical anymore for the termination of the processing.

Reply:
The watch keys will get closed when the service is closed...

Reply:
I've pushed a change to remove the cleanup method entirely, and just close the watch service

Reply:
I'm sorry, but I realized there's one more problem with removing the watch key cleanup, they really should be cleaned up and reset to null in the finally. The service can start and stop number of times, I don't think we should be comparing watch keys from prior watch service on the next start.

Reply:
Good point, I'll null them out make sure they're cleared too